### PR TITLE
Uplift AttnMatmul to support GroupAttnMatmul

### PR DESCRIPTION
**Pull request by TT-BrianLiu**

### CODE REVIEW

File: tests/tt_eager/python_api_testing/unit_testing/test_attn_matmul.py [hunk 1]
```diff
@@ -4,21 +4,13 @@
 
 import pytest
 import torch
+import pytest
```
Comment (aliuTT) [inline tests/tt_eager/python_api_testing/unit_testing/test_attn_matmul.py:7]: remove?

File: tt_eager/tt_dnn/op_library/transformer_tms/kernels/compute/transformer_group_attn_matmul.cpp [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+                    tile_regs_wait();
+                    pack_tile(0, cb_intermed0);
+                    tile_regs_release();
+                    cb_push_back(cb_intermed0, onetile);
+
+                    // untilize tile and write to CB::c_intermed1
+                    unpack_reconfig_data_format_srca(cb_in1, cb_intermed0);
+                    cb_wait_front(cb_intermed0, onetile);
+                    untilize_init_short(cb_intermed0);
+                    cb_reserve_back(cb_intermed1, 1);
```
Comment (aliuTT) [inline tt_eager/tt_dnn/op_library/transformer_tms/kernels/compute/transformer_group_attn_matmul.cpp:65]: should this be `onetile`?

File: tt_eager/tt_dnn/op_library/transformer_tms/multi_core_group_attn_matmul/multi_core_group_attn_matmul.cpp [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+        in1_mcast_sender_noc_y[core_idx_y] = device->worker_core_from_logical_core({0, core_idx_y}).y;
+    }
+
+    // Set up CBs
+    const bool in0_is_sharded = a.is_sharded();
+    const bool in1_is_sharded = b.is_sharded();
+    const bool output_is_sharded = output.is_sharded();
+
+    // CB for in0 (ie. q_heads)
+    uint32_t src0_cb_index = CB::c_in0;
```
Comment (aliuTT) [inline tt_eager/tt_dnn/op_library/transformer_tms/multi_core_group_attn_matmul/multi_core_group_attn_matmul.cpp:75]: `uint32_t src0_cb_index = (uint32_t)CB::c_in0;`
explicit conversion always better :)

File: tt_eager/tt_dnn/op_library/transformer_tms/multi_core_group_attn_matmul/multi_core_group_attn_matmul.cpp [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+		.set_page_size(cb_intermed1_index, interm_single_tile_size);
+    auto cb_interm1 = tt_metal::CreateCircularBuffer(program, all_device_cores, cb_interm1_config);
+
+    uint32_t cb_intermed2_index = CB::c_intermed2;
+    tt_metal::CircularBufferConfig cb_interm2_config = tt_metal::CircularBufferConfig(1 * interm_single_tile_size, {{cb_intermed2_index, interm_data_format}})
+		.set_page_size(cb_intermed2_index, interm_single_tile_size);
+    auto cb_interm2 = tt_metal::CreateCircularBuffer(program, all_device_cores, cb_interm2_config);
+
+    // CB for output (if sharded, full num tiles per core)
+    uint32_t output_cb_index = CB::c_out0; // output operands start at index 16
```
Comment (aliuTT) [inline tt_eager/tt_dnn/op_library/transformer_tms/multi_core_group_attn_matmul/multi_core_group_attn_matmul.cpp:130]: `uint32_t src0_cb_index = (uint32_t)CB::c_in0;`
explicit conversion always better :)

File: tt_eager/tt_dnn/op_library/transformer_tms/multi_core_group_attn_matmul/multi_core_group_attn_matmul.cpp [hunk 5]
```diff
@@ -0,0 +1,448 @@
...
+    const bool src0_is_dram = src0_buffer->buffer_type() == tt_metal::BufferType::DRAM ? 1 : 0;
+    const bool src1_is_dram = src1_buffer->buffer_type() == tt_metal::BufferType::DRAM ? 1 : 0;
+    std::vector<uint32_t> reader_compile_time_args = {
+        (uint32_t) src0_is_dram,
+        (uint32_t) src1_is_dram,
+        (uint32_t) transpose_hw_bool,
+        (uint32_t) row_major,
+    };
+
+    bool dst_is_dram = dst_buffer->buffer_type() == tt_metal::BufferType::DRAM ? 1 : 0;
```
Comment (aliuTT) [inline tt_eager/tt_dnn/op_library/transformer_tms/multi_core_group_attn_matmul/multi_core_group_attn_matmul.cpp:153]: why not use "true" and "false"

File: tt_eager/tt_dnn/op_library/transformer_tms/transformer_tms.cpp [hunk 6]
```diff
@@ -4,6 +4,7 @@
 
 #include "tt_dnn/op_library/transformer_tms/transformer_tms.hpp"
 #include "tt_metal/tools/profiler/op_profiler.hpp"
+#include "tt_dnn/op_library/work_split.hpp"
```
Comment (aliuTT) [inline tt_eager/tt_dnn/op_library/transformer_tms/transformer_tms.cpp:7]: obviously no clang-format

File: tt_eager/tt_dnn/op_library/transformer_tms/transformer_tms.cpp [hunk 7]
```diff
@@ -236,6 +234,167 @@ const operation::Hash AttnMatmul::compute_program_hash(const std::vector<Tensor>
...
 
+
+void GroupAttnMatmul::validate(const std::vector<Tensor>& input_tensors) const {
+    // input_a: [q_len, q_heads, batch, head_dim]
+    // input_b: [batch, kv_heads, head_dim, kv_len]
+    // intermediate: [q_heads, batch, batch, kv_len]
+    // output: [q_len, q_heads, batch, kv_len]
+
+    TT_FATAL(input_tensors.size() == 2);
+    const auto& input_tensor_a = input_tensors.at(0);
```
Comment (aliuTT) [inline tt_eager/tt_dnn/op_library/transformer_tms/transformer_tms.cpp:245]: do you expect value to be populated? if so use `[0]`

File: tt_eager/tt_dnn/op_library/transformer_tms/multi_core_group_attn_matmul/multi_core_group_attn_matmul.cpp [hunk 8]
```diff
@@ -0,0 +1,448 @@
...
+            set_runtime_args
+        ]
+    (
+        const void* operation,
+        Program& program,
+        const std::vector<Tensor>& input_tensors,
+        const std::vector<std::optional<const Tensor>>&,
+        const std::vector<Tensor>& output_tensors
+    ) {
+        const auto& output_tensor = output_tensors.size() == 1 ? output_tensors.at(0) : input_tensors.at(0);
```
Comment (tt-aho) [inline tt_eager/tt_dnn/op_library/transformer_tms/multi_core_group_attn_matmul/multi_core_group_attn_matmul.cpp:437]: This isn't an in place op, so should always be from output_tensors.at(0)

File: tests/tt_eager/python_api_testing/unit_testing/test_attn_matmul.py (hunk 1 above)
Comment (TT-BrianLiu) [reply tests/tt_eager/python_api_testing/unit_testing/test_attn_matmul.py:7]: Thanks i got trolled by Austin

File: tt_eager/tt_dnn/op_library/transformer_tms/multi_core_group_attn_matmul/multi_core_group_attn_matmul.cpp [hunk 9]
```diff
@@ -0,0 +1,448 @@
...
+        reader_kernel_defines["IN0_SHARDED"] = "1";
+    }
+    if (in1_is_sharded) {
+        reader_kernel_defines["IN1_SHARDED"] = "1";
+    }
+    if (output_is_sharded) {
+        writer_kernel_defines["OUT_SHARDED"] = "1";
+    }
+
+    tt_metal::NOC reader_noc = tt_metal::detail::GetPreferredNOCForDRAMRead(tt::Cluster::instance().arch()); // Default is NOC_1
```
Comment (tt-aho) [inline tt_eager/tt_dnn/op_library/transformer_tms/multi_core_group_attn_matmul/multi_core_group_attn_matmul.cpp:171]: Just use device->arch() here?

File: tt_eager/tt_dnn/op_library/transformer_tms/transformer_tms.cpp [hunk 10]
```diff
@@ -236,6 +234,167 @@ const operation::Hash AttnMatmul::compute_program_hash(const std::vector<Tensor>
...
+        TT_FATAL(shard_shape[0] == input_tensor_a.shape()[2]);
+        TT_FATAL(shard_shape[1] == input_tensor_a.shape()[3]);
+    }
+    if (input_tensor_b.is_sharded()) {
+        TT_FATAL(input_tensor_b.memory_config().memory_layout == TensorMemoryLayout::HEIGHT_SHARDED);
+        TT_FATAL(input_tensor_b.shard_spec().value().orientation == shard_orientation, "Any sharded memory configs must have the same shard orientation as one another!");
+        auto shard_shape = input_tensor_b.shard_spec().value().shape;
+        TT_FATAL(shard_shape[0] == input_tensor_b.shape()[1] * input_tensor_b.shape()[2]);
+        TT_FATAL(shard_shape[1] == input_tensor_b.shape()[3]);
+        TT_FATAL(input_tensor_b.shape()[0] == 32, "Only batch 32 is supported for KV sharded!");
```
Comment (tt-aho) [inline tt_eager/tt_dnn/op_library/transformer_tms/transformer_tms.cpp:279]: 😢

File: tt_eager/tt_dnn/op_library/transformer_tms/transformer_tms.cpp [hunk 11]
```diff
@@ -236,6 +234,167 @@ const operation::Hash AttnMatmul::compute_program_hash(const std::vector<Tensor>
...
+    TT_FATAL((ashape[0] == 1), "Input q_len must be 1!");
+    TT_FATAL((ashape[1] % bshape[1] == 0), "Number of q_heads must be divisible by kv_heads!");
+    TT_FATAL((ashape[2] == bshape[0]), "Num of users must match!");
+
+    const auto num_cores_used = std::max(ashape[1], TILE_HEIGHT);  // Need at least 32 cores for mcasting KV heads
+    TT_FATAL((num_cores_used <= this->compute_with_storage_grid_size.x * this->compute_with_storage_grid_size.y), "Compute grid size is too small for group attention matmul! For now, we require at most 1 q_heads per core.");
+
+
+    // Any sharded memory configs must be HEIGHT_SHARDED and have the same orientation
+    ShardOrientation shard_orientation = row_major ? ShardOrientation::ROW_MAJOR : ShardOrientation::COL_MAJOR;
```
Comment (tt-aho) [inline tt_eager/tt_dnn/op_library/transformer_tms/transformer_tms.cpp:265]: Use this->row_major

File: tt_eager/tt_dnn/op_library/transformer_tms/transformer_tms.cpp (hunk 10 above)
Comment (TT-BrianLiu) [reply tt_eager/tt_dnn/op_library/transformer_tms/transformer_tms.cpp:279]: trivial

**Comment by TT-BrianLiu** [comment]
Passing pipelines:
- fast dispatch post-commit: https://github.com/tenstorrent-metal/tt-metal/actions/runs/7661046112/job/20879649184
- slow dispatch post-commit: https://github.com/tenstorrent-metal/tt-metal/actions/runs/7661086148

**Review by arakhmati — APPROVED** [review]

**Review by tt-aho — APPROVED** [review]

**Pull request merged**